mlflow/xgboost.py

```diff
@@ -184,3 +192,89 @@ def __init__(self, xgb_model):
     def predict(self, dataframe):
         import xgboost as xgb
         return self.xgb_model.predict(xgb.DMatrix(dataframe))
+
+
+@experimental
+def autolog(importance_types=['weight']):  # pylint: disable=W0102
+    """
+    Enables autologging from XGBoost to MLflow. Logs the following.
+
+    - parameters specified in `xgboost.train`_.
+    - metrics on each iteration (if ``evals`` specified).
+    - ``best_iteration``, ``best_ntree_limit``, and ``best_score``
+      (if ``early_stopping_rounds`` specified).
+    - feature importance.
+    - trained model.
+
+    :param importance_types: importance types to log.
+
+    """
+    import xgboost
+
+    @gorilla.patch(xgboost)
+    def train(*args, **kwargs):
+
+        def mlflow_callback(env):
+            """
+            Callback for auto-logging metrics on each iteration.
```

Comment:
How quick are the XGBoost iterations? We may end up wanting to log these metrics asynchronously if there are many iterations so that the API request overhead doesn't slow down training, as in [TensorFlow autologging](https://github.com/mlflow/mlflow/blob/master/mlflow/tensorflow.py#L53). However we could add the asynchronous logging in a follow-up & start with something synchronous

Reply:
The iterations speed depends on the # of records and features. I'll take care of the asynchronous logging in a follow-up PR if necessary.